### ISSUE TITLE
Improve search relevance on NuGet.org

### ISSUE THREAD

**Issue by anangaur** [NuGet/NuGetGallery#4124]
1. Search for Microsoft Bot Connector vs. Microsoft Bot Connector - obvious that search is broken here.
Reference: https://twitter.com/adpedley/status/874460132725235712
2. https://github.com/NuGet/NuGetGallery/issues/2789: Support partial search terms in search: E.g. "Microsoft.Bot.Con". Search has no results:
![image](https://user-images.githubusercontent.com/14800916/27108912-a6b6b0be-5054-11e7-9111-0771a7f40608.png)
3. https://github.com/NuGet/NuGetGallery/issues/782: Provide gestures for search filters
4. Enable auto complete in the search box.

**Comment by anangaur** [comment]
Added a couple more.

**Comment by unniravindranathan** [comment]
Terms to test: "UWP"

**Comment by skendrot** [comment]
Search for [uwp](https://www.nuget.org/packages?q=uwp). Results are not as expected. Not sure what I expect, but not what is given where the UWP community toolkit (with 50k downloads) is below items with 500 downloads

**Comment by skofman1** [comment]
Hi @skendrot , NuGet.org search results are heavily based on popularity. Unfortunately, the current popularity calculation is based on partial information, and is essential incorrect: #3778

We are investigating how to fix this issue and provide accurate search results. You can track the attached issue for updates.

**Comment by onovotny** [comment]
Even for popularity, it's not quite accurate either....

Search for Rx has Rx.NET on page 6 of the results…which no one will ever see.
https://preview.nuget.org/packages?q=rx&page=6

**Comment by svick** [comment]
@skofman1 I think they should be based on popularity much more. For example, [when I search for "csharp"](https://preview.nuget.org/packages?q=csharp), hoping to find `Microsoft.CodeAnalysis.CSharp`, some of the results are:

| Position | Package | Downloads |
|-|-|-|
| 2 | `iTextSharp` | 1 800 000 |
| 3 | `CefSharp.WinForms` | 180 000 |
| 6 | `Tarantool.CSharp` | 2 500 |
| 7 | `Hallmanac.Funqy.CSharp` | 30 |
| 8 | `Microsoft.CodeAnalysis.Scripting.CSharp` | 36 000 |
| 9 | `aliyun.oss.csharp.sdk.netstandard1.6` | 0 |
| 10 | `web-csharp-001-core` | 330 |
| 13 | `Microsoft.CodeAnalysis.CSharp` | 6 000 000 |

I don't see any reason why a package that nobody ever downloaded should be higher than a package that was downloaded six million times for this search term.

**Comment by anangaur** [comment]
Linking other search issue: https://github.com/NuGet/NuGetGallery/issues/2405

**Comment by anangaur** [comment]
@skofman1 perhaps we need a search category for "New and upcoming"/"Rising stars" packages apart from "Popular packages"

**Comment by hudo** [comment]
Would like to add another example: Entity Framework Core. Its similar to EF6, which is always 1st in the list.

**Comment by mungojam** [comment]
Is there any tracking, so if somebody searches for term x and then downloads package y, perhaps that should increase the prominence of package y for term x. Maybe that's too easy to game though.

**Comment by anangaur** [comment]
More reports on broken search. In this case its related to metadata relevancy (seems like the tags are given more than required weightage than required?)
https://twitter.com/AnthonySteele/status/901018032969068550

**Comment by agr** [comment]
Related: https://github.com/NuGet/NuGetGallery/issues/4687

**Comment by anangaur** [comment]
++ https://github.com/NuGet/NuGetGallery/issues/4731

**Comment by anangaur** [comment]
Similar:
Search for "Microsoft graph" vs. "Microsoft.Graph"

**Comment by enyim** [comment]
Being based on popularity still does not explain why do we have tens of unrelated packages before the one that actually has the matching id? 

![screen shot 2017-10-20 at 12 13 06 am](https://user-images.githubusercontent.com/78400/31796860-af02f656-b52b-11e7-9bf5-fc929884ff58.png)

**Comment by anangaur** [comment]
@enyim Could be same as mentioned my previous [comment](https://github.com/NuGet/NuGetGallery/issues/4124#issuecomment-333949112). the spaces do not fare well in the NuGet.org search currently -- something that should fix.

**Comment by anangaur** [comment]
From https://github.com/NuGet/Home/issues/6179
Search for `microsoft.aspnetcore.static` -- No Results
Need to search for the full ID: `microsoft.aspnetcore.staticfiles`

**Comment by jskeet** [comment]
Not sure whether this is the right place to raise this, but the popularity sensitivity does seem to be pretty extreme.

Consider [a search for "google cloud storage"](https://www.nuget.org/packages?q=google+cloud+storage). I'd expect a package that has all three search terms *in the title* (and as tags) to rank higher than one which only has "storage" in the title, and doesn't have either "google" or "cloud" in the title, text or tags... but supposedly Windows.AzureStorage is a better match than Google.Cloud.Storage.V1.

I realize search ranking is difficult and often subjective, but this feels like a pretty clear case where popularity is contributing *way* too much to ranking.

**Comment by onovotny** [comment]
Finding another NuGet search issue:
https://www.nuget.org/packages?q=bouncycastle

Looking at that URL, the main BouncyCastle library is properly at the top. However, my Portable.BouncyCastle library, with 1.2M downloads, is “below the fold,” and below “Portable.BouncyCastle-Signed” with only 50k downloads. My library also has been updated far more recently than the other one (3m ago vs >2yrs ago).

I would expect my library to be fourth on that search list, at the very least.

**Comment by anangaur** [comment]
From https://github.com/NuGet/NuGetGallery/issues/5332:
Search for 'pinvoke' on nuget.org
it will give you packages in random order where packages with as low as 120 downloads are in top list.

Now, search for 'pinvoke author:andrew',
it will give andrew arnott's pinvoke libraries which has hundreds of thousands downloads. 
Should first search query not have listed these packages in top list?

**Comment by pebezo** [comment]
The ordering is terrible. If I search for `linq2db` the first entry is a "satellite" package that has the search term at the end of the name with 5K downloads and last updated over two years ago. The next two are similar, but with even fewer downloads (less than 200). 

The right match is the **fouth item** with an exact (100%) match and 95K downloads updated less than 24 hours. How in the world are the first three entries above this one?!

**Comment by RichiCoder1** [comment]
Constantly run into this, especially when I'm doing odd searches like looking for all packages in a "namespace" (aka `AspNetCore.Security.`) and not getting any (or terribly ordered results). Also while searching like [ananguar's](https://github.com/NuGet/NuGetGallery/issues/4124#issuecomment-359896832) where I search for a relatively innocuous tag and get a bunch of irrelevant packages before the one I was actually interested (in an order that doesn't on the surface make much sense).

**Comment by anangaur** [comment]
From Abhijeet on mail:
When I search using following text:
1)	cosmosdb (all lower-case)
a.	https://www.nuget.org/packages?q=documentdb
b.	The package was showed in the middle of the page.
2)	CosmosDb (camelCase)
a.	https://www.nuget.org/packages?q=DocumentDb
b.	The search performed worst in this case (was showed as the last package in the search results)

Does “case” makes a difference in search results?

But what made me ask this question is, the top downloaded package is this with 2+ million downloads and that package is not shown on the top.
![image](https://user-images.githubusercontent.com/14800916/37362105-a4234da6-26b1-11e8-932f-c1b0c8b1964a.png)

**Comment by anangaur** [comment]
https://github.com/NuGet/NuGetGallery/issues/4687#issuecomment-329686422

**Comment by anangaur** [comment]
PM UI search issue:
https://twitter.com/eiivind/status/1035226721937776640

**Comment by anangaur** [comment]
https://twitter.com/dotMorten/status/1039978503519723520:

..nuget searching and result ordering absolutely sucks. Also impossible to filter out packages that doesn't support the TFMs you care about:
![image](https://user-images.githubusercontent.com/14800916/45452286-6d4c5680-b692-11e8-998c-3a9b225219eb.png)
---
![image](https://user-images.githubusercontent.com/14800916/45452296-73423780-b692-11e8-9e2e-f471cea451cb.png)

**Comment by Jay-Rad** [comment]
Yes, please.  I love .NET.  But frankly, NuGet is worse than NPM.

**Comment by anangaur** [comment]
Hi @Jay-Rad, Is your comment just about search or more than that? I would love to discuss any other issues regarding NuGet with you.

**Comment by Jay-Rad** [comment]
It's primarily about the search. At least, within Visual Studio. I don't have any experience with the website or CLI tool.

It seems I can never find anything unless I know the full package name. For example, I needed the Entity Framework Core package from Microsoft. Trying to search using keywords produces hundreds of completely irrelevant and barely-downloaded packages.

That's my biggest problem. I can't find what I need. I have to open other projects that have the package installed so I can remember the full name.

Hope that feedback helps!

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by skofman1** [comment]
@davkean reported:
I cannot figure out the order in which gallery returns search results. If I search for "FxCop", it's returning old packages that haven't been updated in 4 years and have very little downloads ahead of a package that has over 1 million downloads and was last updated 2 weeks ago.

![image](https://user-images.githubusercontent.com/1103906/51583193-83f99880-1f23-11e9-8ce4-223a4706e022.png)

Google and Bing both return results I expect:

https://www.google.com.au/search?source=hp&ei=re1HXOCMOcqP9QOHvaXwBA&q=FxCop+site%3Anuget.org&btnK=Google+Search&oq=FxCop+site%3Anuget.org
https://www.bing.com/search?q=FxCop+site%3Anuget.org&qs=n&form=QBLH&sp=-1&pq=fxcop+site%3A&sc=0-11&sk=&cvid=D4C623BA1F304A9C954F019ED3874526


Note, we since delisted those packages so while this doesn't repro now, the bug/results are still broken.

**Comment by anangaur** [comment]
Another report from a [twitter thread](https://twitter.com/dotMorten/status/1113664725638758400) (may be already captured above)

![image](https://user-images.githubusercontent.com/14800916/55584749-d69cc100-56d9-11e9-999e-6378860c7188.png)

**Comment by TylerBrinkley** [comment]
How is this such a low priority for the NuGet team? This has been a problem forever and yet there hasn't been even a hint of improvements coming.

**Comment by skofman1** [comment]
@TylerBrinkley , we are working on improving search since September: https://github.com/NuGet/NuGetGallery/issues/6371

**Comment by anangaur** [comment]
Related issue from #4687:
Example: I want to find Microsoft.AspNetCore.Mvc:
1. Search for: `microsoft aspnetcore mvc`: https://www.nuget.org/packages?q=microsoft+aspnetcore+mvc
   * The package I want is at the top of page 3: https://www.nuget.org/packages?page=3&q=microsoft%20aspnetcore%20mvc
2. Search for: `microsoft.aspnetcore.mvc`: https://www.nuget.org/packages?q=microsoft.aspnetcore.mvc
   * The package I want is the first match.

**Comment by anangaur** [comment]
From https://developercommunity.visualstudio.com/content/idea/505919/nuget-package-serach-by-a-part-of-the-package-name.html:

Let's say I want to install the Nnuget package: "Microsoft.Owin.Host.SystemWeb", to do that I am going to the Nuget Package Manager and start typing: "Microsoft.Owin.Host.Sys" in the mean time the search goes ahead and says there are no packages. It will be good for the Nuget Package Manager to do search by parts of the package name not the full package name(as a google search would do).

**Comment by anangaur** [comment]
From https://github.com/NuGet/NuGetGallery/issues/2789: 

This is in re: this twitter thread: https://twitter.com/sbohlen/status/571267522571468800

Essentially this comes down to the following issue, which (depending on your POV) either merely hamper discoverability or indicate mostly-broken search functionality:
• Partial matches of search terms names are not considered

Unless you know EXACTLY the package name for which you’re searching, it’s nearly impossible to use search to progressively refine/discover the package(s) you’re seeking.

Example: a search for "nhibernate" finds the NHibernate package:
http://www.nuget.org/packages?q=nhibernate

However, a search for "nhib" does not:
http://www.nuget.org/packages?q=nhib

**Comment by skofman1** [comment]
From @Anthony-Breneliere https://github.com/NuGet/NuGetGallery/issues/7081
> When I execute the command:
> 
> nuget list microsoft dependencyinjection,
> 
> I expect to have the exact package name and version I need:
> Microsoft.Extensions.DependencyInjection 2.20
> and nothing else.
> 
> But instead I have a thousand of packages that just have microsoft in their description.
> 
> The only reliable way I found to get my package is using a Google search:
> https://www.google.com/search?q=nuget+microsoft+dependencyinjection

**Comment by AmadeusW** [comment]
Searching for **PInvoke** returns a package with 5k downloads on first page
and package with 1M downloads on third page. Why?

Does @AArnott need to get a blue checkmark for his packages to boost their placement?
Is it because his version is 0.x, and another is 2.x?
Is it because he did not use tags?

![image](https://user-images.githubusercontent.com/1673956/56376565-0a3d1800-61bd-11e9-9434-ef89cf753637.png)

![image](https://user-images.githubusercontent.com/1673956/56376540-fc879280-61bc-11e9-81b8-7a4db8598de7.png)

NuGet should give me relevant results (Andrew's package) even when author did not go an extra step to get a checkmark and use tags.

Also, it might be better to intersperse results from both authors rather than creating a page-long wall of packages, to show the breadth of options.

**Comment by skofman1** [comment]
@AmadeusW , thanks for the feedback! We will add this example to our test matrix for the new search relevancy algorithm!

**Comment by joelverhagen** [comment]
Mentioned by @AccuPhoenix01:

> Looking for Nuget packages is hard because I need to know the full or exact name of a package. Partial string search seems to be broken as to how I’d expect it to work.
> 
> Steps to reproduce (in VS):
> 1. Open VS project
> 2. Right click on Packages in left sidebar
> 3. Choose “Add Packages”
> 4. In Search area of Add Packages window, look for:
> “mpart” …. looking for mParticle, does not return anything
> “mparticle” … will return mParticle packages
> “appsf” … looking for AppsFlyer, does not return anything
> “appsflyer” … will return packages
> 
> Expected result: Entering any partial string of package name will match.
> Such as “mpart” WILL return all 3 of the mParticle packages.
> 
> This applies to both Web and Visual Studio plug in.

**Comment by anas2204** [comment]
Another one: Search for "FluentAssertion" and the main library isn't even there on the page
![image](https://user-images.githubusercontent.com/8613978/57255095-cdb84b80-7007-11e9-9c29-03c3aa95486d.png)

When you finally put an "s" at the end, it appears. Isn't there a way to rank the library downloaded 21M times higher? (or even IN the list!)
![image](https://user-images.githubusercontent.com/8613978/57255171-f8a29f80-7007-11e9-8317-2dd5b96ab500.png)

**Comment by BenBtg** [comment]
This ongoing issue also affects IDEs that use nuget search such as Visual Studio for Mac. 
Search for **Xamarin Forms** and it doesn't even appear in the top 10 results despite having over 8 million downloads. Why can't the results just be ordered by popularity?
<img width="1008" alt="Screenshot 2019-05-08 at 16 22 41" src="https://user-images.githubusercontent.com/70525/57453504-76f87f00-725e-11e9-8734-7ff816cf0e21.png">

**Comment by heng-liu** [comment]
Adding an example: https://developercommunity.visualstudio.com/content/problem/95390/nuget-package-manager-doesnt-show-title.html

**Comment by bgrainger** [comment]
[MySqlConnector](https://www.nuget.org/packages/MySqlConnector/) is the second-most-popular MySQL-related library, with 2.8 million downloads (and regular updates). However, it appears on [page 9 of results for `mysql`](https://www.nuget.org/packages?page=9&q=mysql):

![image](https://user-images.githubusercontent.com/188129/57577886-e5dc0080-7435-11e9-848e-fc800689a29f.png)

I suspect this is because the current search breaks terms on periods, so `ztools.other.MySql.Data`, which only has 9k downloads and was last updated in 2016 is currently the third result, even though it's extremely unlikely to be a relevant for most searches.

**Comment by joelverhagen** [comment]
A couple examples of bad search queries and results are here on this thread
https://twitter.com/leastprivilege/status/1129390319471136768?s=19

**Comment by leastprivilege** [comment]
I would expect IdentityServer4 to be the 1st item in the list

<img width="1902" alt="Screenshot 2019-05-17 16 20 45" src="https://user-images.githubusercontent.com/1454075/57979167-07328300-7a1a-11e9-99a0-9294eabc68bb.png">

**Comment by leastprivilege** [comment]
...same goes for IdentityModel

<img width="1902" alt="Screenshot 2019-05-19 09 40 20" src="https://user-images.githubusercontent.com/1454075/57979175-26c9ab80-7a1a-11e9-882c-cd19c17c0aff.png">

**Comment by SQL-MisterMagoo** [comment]
I don't have a specific search term to add to the list, but would love to see filtering and sorting.

Possible filters/sorts would be : 

- Dependencies
- Last updated
- Author
- License
- Download total
- Downloads in last 90 days
- Prerelease flag
- Tags

Basically any information you have decided would be useful to show me on the package view would potentially be useful in a search.

**Comment by Advanium** [comment]
Are there any operators to remove specific terms from the result? Such as "serial number -port" because I'm not interested in libraries interacting with serial ports.

**Comment by joelverhagen** [comment]
@Advanium, great idea! Do you think you could file a separate issue for this specific case? We will likely tackle search relevancy/ranking separately from search syntax changes.

**Comment by bgrainger** [comment]
Another example: "MySqlConnector" is listed as the 8th result for the query `MySqlConnector`, even though it's an exact match and has the most downloads of any result.

![image](https://user-images.githubusercontent.com/188129/60125019-bfe99180-973f-11e9-90ac-b2b594455a82.png)

**Comment by loic-sharma** [comment]
@bgrainger We’re working on a new search service which strongly favors exact matches on the package id. Stay tuned! :)

**Comment by 304NotModified** [comment]
@karann-msft 

When search for "log", NLog isn't in the first page. It was first on item 2. I don't get it. The name almost matches and it has the tag "log"

https://www.nuget.org/packages?q=log

**Comment by 304NotModified** [comment]
@loic-sharma thanks, that makes senses.

**Comment by ThomasArdal** [comment]
When searching for `elmah.io` the search result shows the elmah.io package in the first place (correct). The second place is `System.IO` which doesn't make a lot of sense in my head. I need to scroll to the 5th place to find the `Elmah.Io.Client` package.

When searching for `serilog.sinks.elmahio`, the actual package with that id is on the 6th place. Even though a package like `Serilog.Sinks.Console` is more popular, it is not more relevant. I think that when inputting an exact match for a package id, the found package should always be first.

**Comment by ThomasArdal** [comment]
> we're concerned about typosquatting attacks: evil Bob may upload a malicious package named "NLoog" and hope that people accidentally choose that package when they search "NLog".

I see the point, but reserved prefixes will help with "attacks" like this 👍

**Comment by 304NotModified** [comment]
> I think that when inputting an exact match for a package id, the found package should always be first.

Agree on that (but only if there is a really exact match)

**Comment by ThomasArdal** [comment]
> Agree on that (but only if there is a really exact match)

Sure. A search for `Nmog` should not match `NLog`, but both `nlog`, `NLog` and `Nlog` should 👍

**Comment by Jericho** [comment]
Anybody has reported issues since the new search went live when paging through search result using NuGet.Protocol.Core.v3 version 4.3.0-beta1-2418? Specifically, I seeing duplicates. For example a given package is returned on page 2 and again on page 3. I will raise a new issue and prove C# code sample to reproduce but thought I would mention it here.

**Comment by joelverhagen** [comment]
Thanks for @Jericho. Yes, please file a separate issue with repro for non-search relevancy issues.

**Comment by Jericho** [comment]
@joelverhagen Issue #7494 raised

**Comment by fileman** [comment]
If you search Microsoft.AspNetCore.Components.Authorization in Visual Studio or nuget.org ([at page 7](https://www.nuget.org/packages?page=7&q=Microsoft.AspNetCore.Components.Authorization))

**Comment by jedimanu-ux** [comment]
The VSTest platform installer runs this command:
nuget.exe list "Microsoft.TestPlatform" -PreRelease -Source https://api.nuget.org/v3/index.json
which takes ~10 minutes to run. 
Before it took a few seconds. Any idea of an workaround? I'm using nuget.exe 5.2

**Comment by joelverhagen** [comment]
@jedimanu-ux: if you want to list just that package, you can modify the search query to be "package:Microsoft.TestPlatform". `nuget.exe list` essentially uses the same logic as search on nuget.org so you can test your query like this:

https://www.nuget.org/packages?q=Microsoft.TestPlatform
vs.
https://www.nuget.org/packages?q=packageid%3AMicrosoft.TestPlatform

**Comment by raffaeler** [comment]
Is this thread still relevant for improving the search?
I tremendously miss the ability to search dotnet CLI tools only.

**Comment by joelverhagen** [comment]
Hey @raffaeler, we are working on that feature right now in conjunction with the .NET CLI team. The nuget.org work is tracked here: https://github.com/NuGet/NuGetGallery/issues/7594. Our issue is just about making package type filterable for the first time. The .NET CLI issue is here: https://github.com/dotnet/sdk/issues/3808

**Comment by 304NotModified** [comment]
@loic-sharma 

this isn't fixed yet, isn't? https://github.com/NuGet/NuGetGallery/issues/4124#issuecomment-525036144

**Comment by loic-sharma** [comment]
Hi @304NotModified, we've made progress on improving our search results. Some recent changes include:

1. **Boost results that match all search tokens** (https://github.com/nuget/nugetgallery/issues/7391): For example, the query `windows.extensions` should prefer results like `System.Windows.Extensions`
1. **Field weight adjustments** (https://github.com/NuGet/NuGetGallery/issues/7381): As part of this work, we boosted the affect of tag matches
1. **Prefix matching** (https://github.com/nuget/nugetgallery/issues/7128): this lets searches like `nl` match results like `NLog`

These changes have had measurable impact in metrics we track, like bounce rate and search click index on nuget.org, as well as the installation rate in Visual Studio per search session.

That being said, we still have work to do! In [my comment here](https://github.com/NuGet/NuGetGallery/issues/4124#issuecomment-525036144), I had mentioned that we expected that `NLog` would become the first result for the `log` query once we boosted tags. After further investigation, we determined that we were boosting tags a little too strongly, which hurt more queries than it helped. We still plan to improve the `log` and `unit` queries by improving tokenization on initialisms (tracked by https://github.com/NuGet/NuGetGallery/issues/6964).

As always, please let us know if you have any feedback or if you notice other queries that return poor results! 😊

**Comment by jorgensigvardsson** [comment]
I believe the search has been vastly improved! Great work!

**Comment by dazinator** [comment]
Search for packages based on a tag, for example: https://www.nuget.org/packages?page=8&q=Tags%3A%22seq-app%22

You get a tonne of packages show up that don't have that tag - above ones that do.

**Comment by joelverhagen** [comment]
Hey @dazinator, this particular issue is related to how we currently tokenize tags. We want to fix it and the issue is tracked here: https://github.com/NuGet/NuGetGallery/issues/4878.

**Comment by dazinator** [comment]
@joelverhagen ah ok - sorry I missed that one. Thanks.

**Comment by anangaur** [comment]
IMO, this issue has served its purpose and we have improved search relevancy for most of the examples on this issue. However, search is an area for continuous improvements and we should be looking into addressing those improvements as part of other issues, like:

- Package discovery: Show and search by TFM/package-compatibility #3098 
- Gallery should show which frameworks a package is compatible with #4843
- Improve search: Re-evaluate camel-split strategy #8130 
- Display exact match search result at the top and with clear label #7463 

I would like to close this issue in a week from today, unless we discuss reasons not to. Ideally other issues as mentioned above or listed [here](https://github.com/nuget/nugetgallery/issues?q=is%3Aissue+is%3Aopen+search+sort%3Areactions-%2B1-desc) should be upvoted for future investments in search. 

Btw, we have already started exploring the top item on the search improvements stack: Package discovery: Show and search by TFM/package-compatibility #3098
